src/java.base/share/classes/java/lang/invoke/MethodHandleProxies.java

```diff
@@ -171,64 +218,202 @@ public static <T> T asInterfaceInstance(final Class<T> intfc, final MethodHandle
-        if (methods == null)
-            throw newIllegalArgumentException("not a single-method interface", intfc.getName());
-        final MethodHandle[] vaTargets = new MethodHandle[methods.length];
-        for (int i = 0; i < methods.length; i++) {
-            Method sm = methods[i];
-            MethodType smMT = MethodType.methodType(sm.getReturnType(), sm.getParameterTypes());
-            MethodHandle checkTarget = mh.asType(smMT);  // make throw WMT
-            checkTarget = checkTarget.asType(checkTarget.type().changeReturnType(Object.class));
-            vaTargets[i] = checkTarget.asSpreader(Object[].class, smMT.parameterCount());
+
+        /*
+         * For each interface I define a new hidden class and pass the original and
+         * caller-sensitive method handles to its constructor.
+         *
+         * The hidden classes defined for I is defined in a dynamic module M
+         * which has access to the types referenced by the members of I including
+         * the parameter types, return type and exception types.
+         */
+        ProxyClassInfo pci = PROXY_CLASS_INFOS.get(intfc); // throws IllegalArgumentException
+
+        Object proxy;
+        try {
+            proxy = pci.constructor.invokeExact(pci.originalLookup, target, mh);
+        } catch (Throwable e) {
+            throw uncaughtException(e); // propagates WrongMethodTypeException
         }
-        final InvocationHandler ih = new InvocationHandler() {
-                private Object getArg(String name) {
-                    if ((Object)name == "getWrapperInstanceTarget")  return target;
-                    if ((Object)name == "getWrapperInstanceType")    return intfc;
-                    throw new AssertionError();
-                }
-                public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
-                    for (int i = 0; i < methods.length; i++) {
-                        if (method.equals(methods[i]))
-                            return vaTargets[i].invokeExact(args);
-                    }
-                    if (method.getDeclaringClass() == WrapperInstance.class)
-                        return getArg(method.getName());
-                    if (isObjectMethod(method))
-                        return callObjectMethod(proxy, method, args);
-                    if (isDefaultMethod(method)) {
-                        // no additional access check is performed
-                        return JLRA.invokeDefault(proxy, method, args, null);
-                    }
-                    throw newInternalError("bad proxy method: "+method);
+
+        assert proxy.getClass().getModule().isNamed() : proxy.getClass() + " " + proxy.getClass().getModule();
+        return intfc.cast(proxy);
+    }
+
+    private record LocalMethodInfo(MethodTypeDesc desc, List<ClassDesc> thrown, String fieldName) {}
+
+    private record ProxyClassInfo(MethodHandle constructor, Lookup originalLookup) {}
```

Comment:
`ProxyClassInfo` for interface `I` will be stored in the class value of `I`.  Hence it keeps a strong reference to the generated proxy class for `I` until `I` is unloaded.  The hidden class implicitly becomes a strong link to the defining class loader.    Need to look into keeping it in a weak reference to allow the hidden class to be unloaded independent of the lifetime of the class loader.